Keda + Apache Pulsar Issue

### Report

Keda Operator pod get restarting.

### Expected Behavior

It should work without any issue

### Actual Behavior

keda operator pods get keep restarting. if i delete the apache pulsar sclaer, then it will stop restarting and will work fine.

### Steps to Reproduce the Problem

We are using kubernetes 1.33 on AWS EKS.

APache pulsar 4.0.6

Example scale object

apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: xyz-scaller-pulsar
  namespace: default
  annotations:
    autoscaling.keda.sh/paused: "false"
    scaledobject.keda.sh/transfer-hpa-ownership: "true"
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: xyz-worker
  pollingInterval: 30
  cooldownPeriod: 90
  minReplicaCount: 0
  maxReplicaCount: 10

  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 60   # shorten HPA wait time
  triggers:
    - type: pulsar
      metadata:
        adminURL: http://pulsar-broker.pulsar.svc.cluster.local:8080
        topic: persistent://public/default/articles
        isPartitionedTopic: "false"
        subscription: articles-subscription
        activationMsgBacklogThreshold: "0"
        msgBacklogThreshold: "50"
        authModes: "bearer"
      authenticationRef:
        name: keda-auth-pulsar-credential

### Logs from KEDA operator

```
025-08-27T15:36:39Z	INFO	Initializing Scaling logic according to ScaledObject Specification	{"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "ScaledObject": {"name":"xyz-scaller-pulsar","namespace":"default"}, "namespace": "default", "name": "xyz-scaller-pulsar", "reconcileID": "e3c23390-ddc6-4895-a961-8d85cda32110"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x371bff8]

goroutine 1478 [running]:
github.com/kedacore/keda/v2/pkg/scalers.NewPulsarScaler.func1(0x4291d20?, {0x4000113908?, 0x480c9c6?, 0x7?})
	/workspace/pkg/scalers/pulsar_scaler.go:120 +0x28
net/http.(*Client).checkRedirect(...)
	/usr/local/go/src/net/http/client.go:505
net/http.(*Client).do(0x4004b4d3e0, 0x4004c7d2c0)
	/usr/local/go/src/net/http/client.go:691 +0x598
net/http.(*Client).Do(...)
	/usr/local/go/src/net/http/client.go:590
github.com/kedacore/keda/v2/pkg/scalers.(*pulsarScaler).GetStats(0x4004b3dcb0, {0x4f31bf0, 0x4004c528c0})
	/workspace/pkg/scalers/pulsar_scaler.go:256 +0x200
github.com/kedacore/keda/v2/pkg/scalers.(*pulsarScaler).getMsgBackLog(0x4004b3dcb0, {0x4f31bf0?, 0x4004c528c0?})
	/workspace/pkg/scalers/pulsar_scaler.go:285 +0x30
github.com/kedacore/keda/v2/pkg/scalers.(*pulsarScaler).GetMetricsAndActivity(0x4004b3dcb0, {0x4f31bf0?, 0x4004c528c0?}, {0x4004cb0230, 0x44})
	/workspace/pkg/scalers/pulsar_scaler.go:301 +0x38
github.com/kedacore/keda/v2/pkg/scaling/cache.(*ScalersCache).GetMetricsAndActivityForScaler(0x4004b14600, {0x4f31bf0, 0x4004c528c0}, 0x0, {0x4004cb0230, 0x44})
	/workspace/pkg/scaling/cache/scalers_cache.go:151 +0xe0
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScalerState(0x4000659b20?, {0x4f31bf0, 0x4004c528c0}, {0x4f29190, _}, _, {{0x4004b051a0, 0x20}, {0x4004b4e2b0, 0x7}, ...}, ...)
	/workspace/pkg/scaling/scale_handler.go:758 +0x2bc
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScaledObjectState.func1({0x4f29190?, 0x4004b3dcb0?}, 0x4efe2c8?, {{0x4004b051a0, 0x20}, {0x4004b4e2b0, 0x7}, {0x381b617, 0xc}, 0xb2d05e00, ...}, ...)
	/workspace/pkg/scaling/scale_handler.go:633 +0x90
created by github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScaledObjectState in goroutine 1477
	/workspace/pkg/scaling/scale_handler.go:632 +0x400

```


### KEDA Version

2.16.1

### Kubernetes Version

Other

### Platform

Amazon Web Services

### Scaler Details

Apache Pulsar

### Anything else?

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -97,6 +97,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **General**: Sets hpaName in Status when ScaledObject adopts/finds an existing HPA ([#6336](https://github.com/kedacore/keda/issues/6336))
 - **Cron Scaler**: Fix cron scaler to return zero metric value by default([#6886](https://github.com/kedacore/keda/issues/6886))
 - **Hashicorp Vault**: Fix Vault PKI param using camel case param when making API call to Vault PKI endpoint ([#6864](https://github.com/kedacore/keda/issues/6864))
+- **Pulsar Scaler**: Resolve nil pointer dereference in Pulsar scaler redirect handling ([#7024](https://github.com/kedacore/keda/issues/7024))
 - **RabbitMQ Scaler**: Fix incorrect URL encoding in RabbitMQ vhosts containing %2f ([#6963](https://github.com/kedacore/keda/issues/6963))
 
 ### Deprecations

--- a/pkg/scalers/pulsar_scaler.go
+++ b/pkg/scalers/pulsar_scaler.go
@@ -192,7 +192,7 @@ func NewPulsarScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
 		if pulsarMetadata.pulsarAuth.EnableBearerAuth || pulsarMetadata.pulsarAuth.EnableBasicAuth {
 			// The pulsar broker redirects HTTP calls to other brokers and expects the Authorization header
 			client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
-				if len(via) != 0 && via[0].Response.StatusCode == http.StatusTemporaryRedirect {
+				if len(via) != 0 && via[0] != nil && via[0].Response != nil && via[0].Response.StatusCode == http.StatusTemporaryRedirect {
 					addAuthHeaders(req, pulsarMetadata)
 				}
 				return nil

--- a/pkg/scalers/pulsar_scaler_test.go
+++ b/pkg/scalers/pulsar_scaler_test.go
@@ -3,6 +3,8 @@ package scalers
 import (
 	"context"
 	"fmt"
+	"net/http"
+	"net/http/httptest"
 	"reflect"
 	"strconv"
 	"strings"
@@ -403,5 +405,69 @@ func TestPulsarGetMetric(t *testing.T) {
 		}
 
 		fmt.Printf("%+v\n", metric)
+	}
+}
+
+// Test for nil pointer dereference fix in redirect handling for issue #7024
+func TestPulsarScalerRedirectNilPointerFix(t *testing.T) {
+	// Mock HTTP server that simulates Pulsar broker redirects
+	redirectCount := 0
+	var mockServer *httptest.Server
+
+	mockServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		redirectCount++
+		if redirectCount == 1 {
+			// Send redirect
+			http.Redirect(w, r, mockServer.URL+"/redirected", http.StatusTemporaryRedirect)
+			return
+		}
+		// Return valid response
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		_, err := w.Write([]byte(`{
+			"subscriptions": {
+				"test-subscription": {
+					"msgBacklog": 5
+				}
+			}
+		}`))
+		if err != nil {
+			t.Errorf("Failed to write response: %v", err)
+		}
+	}))
+	defer mockServer.Close()
+
+	config := &scalersconfig.ScalerConfig{
+		TriggerMetadata: map[string]string{
+			"adminURL":     mockServer.URL,
+			"topic":        "persistent://public/default/test-topic",
+			"subscription": "test-subscription",
+			"authModes":    "bearer",
+		},
+		AuthParams: map[string]string{
+			"bearerToken": "test-token",
+		},
+	}
+
+	// Create scaler
+	scaler, err := NewPulsarScaler(config)
+	if err != nil {
+		t.Fatalf("Failed to create scaler: %v", err)
+	}
+
+	// Cast to pulsarScaler to access GetStats method
+	pulsarScaler, ok := scaler.(*pulsarScaler)
+	if !ok {
+		t.Fatal("Expected scaler to be *pulsarScaler type")
+	}
+
+	_, err = pulsarScaler.GetStats(context.Background())
+	if err != nil {
+		t.Logf("GetStats returned error (expected due to test setup): %v", err)
+	}
+
+	// Verify redirect happened
+	if redirectCount < 1 {
+		t.Error("Expected at least one HTTP call to trigger redirect")
 	}
 }

